PEM-4503-managing_certificates copy for trusted CA certificate and keys (#3627)
EnterpriseDB/docs: product_docs/docs/pem/9/certificates/index.mdx

                VALUES(
                    'Renew agent certificate','This job renew the agent certificate', rec_agent.id, now()) 
                RETURNING jobid INTO v_jobid;
                -- create the jobstep for the above created job
                INSERT INTO pem.jobstep(
                    jstjobid, jstname, jstdesc, jstkind, jstcode) 
                VALUES (
                    v_jobid, 'Renew agent certificate','This job step runs to renew the agent certificate.', 'i','renew_agent_certificate') RETURNING jstid INTO v_jobstepid;
                RAISE INFO 'Created a job (#% with step #%) for Agent#% (%)', v_jobid, v_jobstepid, rec_agent.id, rec_agent.description;
            END LOOP;
            CLOSE cur_agent_list;
       END
       $BODY$;
    
    # Execute the above created function
    SELECT pem.create_job_for_renewing_certs();
    ```

7. Restart the PEM backend database server: 
    
   ```shell
   # Running as root
   systemctl restart edb-as-14.service
   ```
    
   To do a sanity check that all the agent-level jobs have completed successfully, go to the PEM agents in the browser tree and check the status of the scheduled tasks.

>    Each agent's new certificates and keys are in `/root/.pem`. If the jobs completed successfully, the PEM server and agents are up and running using new certificates and keys.

Reviewer: If this is correct, I think this flow works better.

> Each agent's new certificates and keys are now in `/root/.pem`.

> To do a sanity check that all the agent-level jobs have completed successfully, go to the PEM agents in the browser tree and check the status of the scheduled tasks. If the jobs completed successfully, the PEM server and agents are up and running using new certificates and keys.

---
title: "Managing SSL certificates"
navTitle: "Managing SSL certificates"
legacyRedirectsGenerated:
  # This list is generated by a script. If you need add entries, use the `legacyRedirects` key.
  - "/edb-docs/d/edb-postgres-enterprise-manager/user-guides/administrators-guide/8.0/managing_certificates.html"
# This file is moved from pem_admin/04_managing_certificates
redirects:
- /pem/latest/managing_certificates/
- /pem/latest/pem_admin/04_managing_certificates/
navigation:
- replacing_ssl_certificates
- regenerating_agent_certificates
---

PEM uses self-signed SSL certificates:

- Webserver(HTTPD)
- PEM agent and backend database server connection

## Webserver certificates

> We recommend an additional layer of SSL security for the web application.

Reviewer 2: Removed

> PEM uses self-signed SSL certificates for the Apache HTTPD server. The self-signed SSL certificates and keys for the Apache HTTPD server are generated during PEM installation.

Reviewer: Is there one HTTPD certificate or more than one? Depending on the answer either make certificates singular in this paragraph or make it plural in the next.
```suggestion
PEM uses self-signed SSL certificates for the Apache HTTPD server. The self-signed SSL certificates and keys for the Apache HTTPD server are generated during PEM installation.
```

Reviewer 3: I believe for Apache httpd (notice the product name is actually lower case, which I neglected to correct previously) there is only one key and one certificate so we could be more specific and use the singular form here.

When referring to agents, there is one certificate per agent (except in the documented case of sharing a single cert) so plural is appropriate.

PEM uses the self-signed SSL certificates for the Apache/HTTPD server. The self-signed SSL certificates and keys for the Apache/HTTPD server get generated during the PEM installation.

If you want to use your own SSL certificate for PEM, you must update the Apache configuration file.

On Linux, you need to open the Apache HTTPD configuration file (`edb-ssl-pem.conf`) in a text editor and be a user with write permission on the file. You must also change the server name and file names in the configuration file to match your certificate files.

Update these two SSL directives in the PEM VirtualHost section:

-   `SSLCertificateFile` is your DigiCert certificate file (for example, your_domain_name.crt).
-   `SSLCertificateKeyFile` is the .key file generated when you created the CSR (for example, your_private.key).

For example, make the following updates:

```shell
SSLEngine on
SSLCertificateFile /path/to/your_domain_name.crt
SSLCertificateKeyFile /path/to/your_private.key
```

> To increase the security, you can replace the HTTPD self-signed SSL certificates with the trusted third-party CA signed certificates in PEM. For more information, see [Replacing HTTPD self-signed SSL certificates](https://www.enterprisedb.com/postgres-tutorials/how-replacing-httpd-self-signed-ssl-certificates-trusted-ca-signed-certificates).

Reviewer 2: This link takes to the full working example for Replacing HTTPD self-signed SSL certificates

## PEM backend database server and agent connection certificates

> PEM, by default, supports secured SSL/TLS connections and acts as its own certificate authority (CA) to generate certificates and keys for the PEM server and agent. The self-signed SSL certificates and keys for the PEM server and agent get generated during the PEM installation. These certificates and keys encrypt the connections between the server and the agent. 

Reviewer 3: ```suggestion
PEM, by default, implements secured SSL/TLS connections between PEM agents and the backend database, and acts as its own certificate authority (CA) to generate certificates and keys for the PEM server and agent. The self-signed SSL certificates and keys for the PEM server and agent are generated during PEM installation. These certificates and keys encrypt the connection from agent to server. In addition, PEM Agents are authenticated using their certificate rather than a password.
```

> To increase the security, you can replace the PEM self-signed SSL certificates and keys with the trusted third-party CA certificates. For more information, see [Trusted third-party CA certificates and keys](/pem/latest/certificates/#trusted-third-party-ca-certificates-and-keys).

Reviewer 3: Same comment as previous occurence

### How PEM self-signed SSL certificates work

The PEM server configuration script generates self-signed SSL certificates and key files for the PEM backend database server. The backend database server uses these certificates and keys to authenticate and encrypt the agent connections. Each certificate has an expiry date. You should regenerate the certificates when it nears expiration. For more information, see [Regenerating server self-signed SSL certificates](/pem/latest/certificates/replacing_ssl_certificates).

The PEM agent connects to the PEM backend database server using the libpq interface, acting as a client of the backend database server. The agent self-signed SSL certificates and keys get generated during the [PEM registration](/pem/latest/registering_agent). PEM agent establishes the connection with the PEM backend database server using the self-signed SSL certificate and key files. 

> Each agent has a unique identifier and the agent certificates and keys have the corresponding identifier mentioned. Each certificate has an expiry date. You should regenerate the certificates when it nears expiration. For more information, see [Regenerating agent self-signed SSL certificates](/pem/latest/certificates/regenerating_agent_certificates).

Reviewer 3: These three paragraphs feel like they are repeating themselves, and repeating things we've said earlier. As a reader I'm not sure if I've failed to notice some subtle extra information or if this is just repetition on the part of the author.

> You can generate a common agent certificate and key file pair for all the PEM agents and use for connecting to PEM backend database server. For more information see [Generate common agent certificate and key pair](/pem/lates/certificates/#generate-a-common-agent-certificate-and-key-pair)

Reviewer 3: This might be worth drawing more attention to and explaining why you might do it.

For more information on using the SSL certificates to connect in Postgres, see [Securing TCP/IP connections with SSL](https://www.postgresql.org/docs/current/ssl-tcp.html).

### Certificate and key files generation

The PEM server generates the certificates and key files in the data directory of the backend database server:

- `ca_certificate.crt`
- `ca_key.key`
- `root.crt`
- `root.crl`
- `server.crt`
- `server.key`

The `ca_certificate.crt` and `ca_key.key` are used during the agent registration process to generate the agent's SSL certificates and key files.

The `root.crt` file is a copy of the `ca_certificate.crt` file. You use the root certificate for the backend database server by setting the `ssl_ca_file` parameter as `root.crt` in the postgresql.conf file.

The `root.crl` has the certificate revocation list (CRL) of digital certificates revoked by the issuing Certificate Authority (CA) before their actual or assigned expiration date.

> The `server.crt` is the signed certificate for the PEM server and the `server.key` is the private key to the certificate. The PEM agent certificates get generated using this server certificate and key files.

Reviewer 3: ```suggestion
The `server.crt` is the signed certificate for the PEM server and the `server.key` is the private key to the certificate. The PEM agent certificates are generated using this server certificate and key files.
```

        ```shell
        export PEM_SERVER_PASSWORD=edb

        /usr/edb/pem/agent/bin/pemworker --register-agent \
        --pem-server 192.168.99.130 \
        --pem-user enterprisedb \
        --pem-port 5444 \
        --pem-agent-user pem_agent_common_user \
        -o agent_ssl_crt= agent.crt \
        -o agent_ssl_key= agent.key
        ```

     b. Enable and start the pemagent services:

        ```shell
        systemctl enable pemagent
        systemctl start pemagent
        ```

   - To replace the agent certificate and key pair with the registered agent:
 
     a. Edit the `agent_user`, `agent_ssl_key`, and `agent_ssl_crt` parameters in `agent.cfg` file of the agent host

        ```shell
        vi /usr/edb/pem/agent/etc/agent.cfg
        # Edit the agent username
        agent_user=pem_agent_common_user
        # Edit the ssl parameters with new certificate and key file location
        agent_ssl_key=<new_location>/agent.key
        agent_ssl_crt=<new_location>/agent.crt
        ```
      
     b. Restart the pemagent service:

        ```shell
        systemctl restart pemagent
        ```

> ## Trusted third-party CA certificates and keys

Reviewer 3: Didn't we provide links to a blog about this earlier on?

Reviewer 2: Blog for replacing httpd certificates and this are for server and agent connection certificates

After obtaining the trusted third-party CA certificates and keys, perform the following steps: 

- Replace PEM self-signed certificates and keys with trusted CA-signed certificates and keys

> - Disable CRL usage by the server if the trusted third-party CA does not provide CRL

Reviewer: Do we need to provide more information on how to do this? The instructions for the previous bullet refers to the following two subsections, correct? If that is the case and this bullet just describes another task you may need to do, add ", on the server and agents" with server and agents linked to the child sections.

```suggestion
- Disable CRL usage by the server if the trusted third-party CA does not provide CRL
```

---

PEM uses self-signed SSL certificates:

- to secure requests to the [web server](#webserver-certificates) (Apache HTTPD) which provides the user interface and REST API.
- to secure and authenticate the [PEM agent connections to the PEM backend database](#pem-backend-database-server-and-agent-connection-certificates).

## Web server certificates

PEM uses self-signed SSL certificates for the Apache HTTPD server. The self-signed SSL certificates and keys for the Apache HTTPD server are generated during PEM installation.

To use your own SSL certificate for PEM, update the Apache HTTP configuration file (`edb-ssl-pem.conf`).

Update these two SSL directives in the PEM VirtualHost section:

-   `SSLCertificateFile` is your certificate file (for example, your_domain_name.crt).
-   `SSLCertificateKeyFile` is the .key file generated when you created the CSR (for example, your_private.key).

For example, make the following updates:

```shell
# Change the server name and file names in the configuration file to 
# match your certificate files.
SSLEngine on
SSLCertificateFile /path/to/your_domain_name.crt
SSLCertificateKeyFile /path/to/your_private.key
```

> To increase security, you can replace the HTTPD self-signed SSL certificates with trusted third-party CA signed certificates in PEM. For more information, see [Replacing HTTPD self-signed SSL certificates](https://www.enterprisedb.com/postgres-tutorials/how-replacing-httpd-self-signed-ssl-certificates-trusted-ca-signed-certificates).

Reviewer 3: ```suggestion
To increase security, you can replace the HTTPD self-signed SSL certificates with trusted third-party CA signed certificates in PEM. For more information, see [Replace the server self-signed SSL certificates with trusted third-party CA certificates](#replace-the-server-self-signed-ssl-certificates-with-trusted-third-party-ca-certificates).
```
Better to link to this doc than drag in a blog article that we're not maintaining

Reviewer 2: The link to the blog provides steps for Replacing httpd SSL certificates. And the link in our doc is for Replacing the Server SSL certificates. Two things are different, as far as I understand. We can consider adding the steps of the blog to our docs so that it can be maintained. We can discuss this and finalise it.

        ```

     b. Enable and start the pemagent services:

        ```shell
        systemctl enable pemagent
        systemctl start pemagent
        ```

   - to replace the agent certificate and key pair with the registered agent.
 
     a. Edit the `agent_user`, `agent_ssl_key`, and `agent_ssl_crt` parameters in `agent.cfg` file of the agent host

        ```shell
        vi /usr/edb/pem/agent/etc/agent.cfg
        # Edit the agent username
        agent_user=pem_agent_common_user
        # Edit the ssl parameters with new certificate and key file location
        agent_ssl_key=<new_location>/agent.key
        agent_ssl_crt=<new_location>/agent.crt
        ```
      
     b. Restart the pemagent service:

        ```shell
        systemctl restart pemagent
        ```

## Trusted third-party CA certificates and keys

You can replace the PEM self-signed SSL certificates and keys with trusted third-party CA certificates and keys. 

After obtaining the trusted third-party CA certificates and keys: 

- Replace PEM self-signed certificates and keys with trusted CA-signed certificates and keys
- Disable CRL usage by the server if the trusted third-party CA does not provide CRL

> ### Replace the server self-signed SSL certificates with the trusted third-party CA certificates

Reviewer 3: ```suggestion
### Replace the server self-signed SSL certificates with trusted third-party CA certificates
```

   ```shell
   openssl x509 -noout -text -in server.crt
   ```

1. Use the new certificate obtained from CA as the `root.crt` file:

   ```shell
   cp ca_certificate.crt root.crt
   ```

1. Copy new `root.crt`, `server.key`, and `server.crt` to the data directory of the backend database server:

   ```shell
   cp root.crt server.key server.crt /var/lib/edb/as<x>/data
   ```

1. Change the owner and permissions of new certificates and key files same as the data directory:

   ```shell
   cd /var/lib/edb/as<x>/data/
   chown enterprisedb server.* root.crt ca_certificate.crt
   chmod 600 server.* root.crt ca_certificate.crt
   ```

    
   !!! Note
       Do not restart the PEM server now, or else all the registered agents will stop working.

1. Replace each PEM agent SSL certificates with the trusted third-party CA certificates. For more information, see this [section](/pem/latest/certificates/#replace-the-agent-self-signed-ssl-certificates-with-the-trusted-third-party-ca-certificates).

1. Restart the PEM server.

   - On Linux:

   ```shell
   # Running as root
   systemctl start edb-as-<x>
   ```

   Restarting the backend database server, restarts the PEM server.

> ### Replace the agent self-signed SSL certificates with the trusted third-party CA certificates

Reviewer 3: ```suggestion
### Replace the agent self-signed SSL certificates with trusted third-party CA certificates
```

> ### Replace the agent self-signed SSL certificates with the trusted third-party CA certificates

Reviewer 3: These instructions only work if you have previously followed the instructions above. I think this is reasonable as most people will want to change both, but we should probably note in this section that you need to have completed the instructions above first - specifically adding the new `root.crt`

Reviewer 3: I discussed this with @asheshv so he should be be able to provide the commands

   ```
   
   Where `<ID>` is the agent identifier.

1. Generate a Certificate Signing Request (CSR) for each agent:

   ```shell
   openssl req -new -key agent<ID>.key -out agent<ID>.csr -subj '/C=IN/ST=MH/L=Pune/O=PEM/CN=agent<ID>'
   ```
   
   Where you replace `<ID>` in `CN` with appropriate agent identifier.

1. Ask your CA to sign the CSR and generate agent certificate for you. 

1. Copy the certificate and key files on respective hosts, where `<ID>` matches the `agent_id` in the `/usr/edb/pem/agent/etc/agent.cfg` file.

1. Change the ownership and permission on the new `agent<ID>.crt` and `agent<ID>.key` file:

    ```shell
    chown root agent<ID>.crt agent<ID>.key
    chmod 600 agent<ID>.crt agent<ID>.key
    ```

1. Take backup of old agent certificate and key file:

   ```shell
   # Running as root
   mkdir root/.pem/certs
   mv root/.pem/agent<ID>.* root/.pem/certs
   ```

1. Replace each agent's certificate and key file with the newly generated files: 
   
   ```shell
   cp agent<ID>.key agent<ID>.crt root/.pem
   ```

7. Restart the PEM agent service.

   - On Linux:

   ```shell
   # Running as root
   systemctl restart pemagent
   ```

   - On Windows:

>      You can use the Services applet to restart the PEM agent. The PEM agent service is named Postgres Enterprise Manager Agent. Select the service name in the Services dialog box and select **Restart the service**.

Reviewer 3: I suggest adding a troubleshooting/test section at the end which covers the following two steps (should be referenced appropriately from instructions above).

- How to validate a certificate against root.crt
- How to test ssl connection from an agent host using psql

Reviewer 3: I have added this in [762a97f](https://github.com/EnterpriseDB/docs/pull/3627/commits/762a97f07ff8074de4fa36a336336bddaa904324) so I'll close this conversation

1. Restart the PEM server.

   - On Linux:

   ```shell
   # Running as root
   systemctl start edb-as-<x>
   ```

   Restarting the backend database server, restarts the PEM server.

### Replace the agent self-signed SSL certificates with trusted third-party CA certificates

1. Use psql to find all the agent identifiers(IDs) we need to replace the SSL certificates:

   ```shell
   psql -U enterprisedb -d pem --no-psqlrc -t -A -c "SELECT id FROM pem.agent WHERE active=true"
   ```

1. After identifying the agents that need key files, generate an `agent<ID>.key` for each agent. To generate the key, execute the command:

   ```shell
   openssl genrsa -out agent<ID>.key 4096 
   ```
   
   Where `<ID>` is the agent identifier.

1. Generate a Certificate Signing Request (CSR) for each agent:

   ```shell
   openssl req -new -key agent<ID>.key -out agent<ID>.csr -subj '/C=IN/ST=MH/L=Pune/O=PEM/CN=agent<ID>'
   ```
   
   Where `-subj` is provided as per your requirements. You replace `<ID>` in `CN` with appropriate agent identifier.

> !!! Note
> If you prefer to use a single certificate for all PEM Agents, rather that one per agent, you should create a common Postgres user
> and supply this username in place of `ID` as detailed in [Generate a common agent certificate and key pair](/pem/latest/certificates/#generate-a-common-agent-certificate-and-key-pair)

Reviewer 3: ```suggestion
    !!! Note
    If you prefer to use a single certificate for all PEM Agents, rather that one per agent, you should create a common Postgres user
    and supply this username in place of `ID` as detailed in [Generate a common agent certificate and key pair](/pem/latest/certificates/#generate-a-common-agent-certificate-and-key-pair)
```

Trying to fix my bad markdown

After obtaining the trusted CA certificates and keys replace the [server](/pem/latest/certificates/#replace-the-self-signed-server-ssl-certificates-with-the-certificates-signed-by-the-trusted-ca) and [agent](/pem/latest/certificates/#replace-the-self-signed-agent-ssl-certificates-with-the-certificates-signed-by-the-trusted-ca) certificates and keys.

### Replace the self-signed server SSL certificates with the certificates signed by the trusted CA

1. Take the backup of old server certificate and key files:

   ```shell
   # Running as root
   mkdir /var/lib/edb/as<x>/data/certs
   cd /var/lib/edb/as<x>/data/
   mv server.* root.* ca_* /var/lib/edb/as<x>/data/certs
   ```

1. Generate a private key for the server:

   ```shell
   openssl genrsa -out server.key 4096 
   ```

1. Generate a Certificate Signing Request (CSR) for the server:

   ```shell
   openssl req -new -key server.key -out server.csr -subj '/C=IN/ST=MH/L=Pune/O=EDB/CN=PEM'
   ```
   
   Where `-subj` is provided as per your requirements. We recommend using the `Hostname/Domain Qualified Full Name` of the PEM server host for `CN`.

>    

Reviewer 4: Not sure why Hostname/Domain Qualified Full Name is in in caps and tagged as code. I think you just want to say:
hostname or domain-qualified full name
with no special tagging

        ```shell
        export PEM_SERVER_PASSWORD=edb

        /usr/edb/pem/agent/bin/pemworker --register-agent \
        --pem-server 192.168.99.130 \
        --pem-user enterprisedb \
        --pem-port 5444 \
        --pem-agent-user pem_agent_common_user \
        -o agent_ssl_crt= agent.crt \
        -o agent_ssl_key= agent.key
        ```

     b. Enable and start the pemagent services:

        ```shell
        systemctl enable pemagent
        systemctl start pemagent
        ```

   - To replace the agent certificate and key pair with the registered agent.
 
     a. Edit the `agent_user`, `agent_ssl_key`, and `agent_ssl_crt` parameters in `agent.cfg` file of the agent host:

        ```shell
        vi /usr/edb/pem/agent/etc/agent.cfg
        # Edit the agent username
        agent_user=pem_agent_common_user
        # Edit the ssl parameters with new certificate and key file location
        agent_ssl_key=<new_location>/agent.key
        agent_ssl_crt=<new_location>/agent.crt
        ```
      
     b. Restart the pemagent service:

        ```shell
        systemctl restart pemagent

>         ```

Reviewer: We will be considering this comment in the next round of changes.

SSLCertificateKeyFile /path/to/your_private.key
```

To increase security, you replace the httpd self-signed SSL certificates with trusted third-party CA signed certificates in PEM. For more information, see [Replacing httpd self-signed SSL certificates](https://www.enterprisedb.com/postgres-tutorials/how-replacing-httpd-self-signed-ssl-certificates-trusted-ca-signed-certificates).

## PEM backend database server and agent connection certificates

PEM, by default, implements secured SSL/TLS connections between PEM agents and the backend database, and acts as its own certificate authority (CA) to generate certificates and keys for the PEM server and agent. The self-signed SSL certificates and keys for the PEM server and agent are generated during PEM installation. These certificates and keys encrypt the connection from agent to server. In addition, PEM agents are authenticated using their certificate rather than a password.

> You replace the PEM self-signed SSL certificates and keys with the trusted third-party CA certificates and keys. For more information, see [Trusted third-party CA certificates and keys](/pem/latest/certificates/#trusted-third-party-ca-certificates-and-keys).

Reviewer 3: I suggest we reinstate 'can' in this sentence to make it clear that this is an optional step.

> You replace the PEM self-signed SSL certificates and keys with the trusted third-party CA certificates and keys. For more information, see [Trusted third-party CA certificates and keys](/pem/latest/certificates/#trusted-third-party-ca-certificates-and-keys).

Reviewer 3: ```suggestion
You can replace the PEM self-signed SSL certificates and keys with the trusted third-party CA certificates and keys. For more information, see [Trusted third-party CA certificates and keys](/pem/latest/certificates/#trusted-third-party-ca-certificates-and-keys).
```

The PEM agent connects to the PEM backend database server using the libpq interface, acting as a client of the backend database server. The agent self-signed SSL certificates and keys get generated during [agent registration](/pem/latest/registering_agent). PEM agent establishes the connection with the PEM backend database server using the self-signed SSL certificate and key files. 

Each agent has a unique identifier and the agent certificates and keys have the corresponding identifier mentioned. Each certificate has an expiry date. You should regenerate the certificates when it nears expiration. For more information, see [Regenerating agent self-signed SSL certificates](/pem/latest/certificates/regenerating_agent_certificates).

> Generate a common agent certificate and key file pair for all the PEM agents and use for connecting to PEM backend database server. For more information, see [Generate common agent certificate and key pair](/pem/lates/certificates/#generate-a-common-agent-certificate-and-key-pair)

Reviewer 3: ```suggestion
If required, it is possible to use the same certificate for all agents rather than one certificate per agent. For more information, see [Generate common agent certificate and key pair](/pem/lates/certificates/#generate-a-common-agent-certificate-and-key-pair)
```

## Generate a common agent certificate and key pair

> Generate one agent certificate and key pair and use them for all the PEM agents.

Reviewer 3: Suggest: "By creating and using a single Postgres user for all PEM agents (rather than one user per agent as is the default) it is possible to use the same certificate for all agents. These instructions explain how to create a user, generate an agent certificate and key pair, and use them for all PEM agents."

> Generate one agent certificate and key pair and use them for all the PEM agents.

Reviewer 3: ```suggestion
By creating and using a single Postgres user for all PEM agents (rather than one user per agent as is the default) it is possible to use the same certificate for all agents. These instructions explain how to create a user, generate an agent certificate and key pair, and use them for all PEM agents.
```
